net/ws.ts

```diff
@@ -0,0 +1,349 @@
+  const hasMask = (frame.mask ? 1 : 0) << 7;
+  if (payloadLength < 126) {
+    header = new Uint8Array([
+      (0b1000 << 4) | frame.opcode,
+      hasMask | payloadLength
+    ]);
+  } else if (payloadLength < 0xffff) {
+    header = new Uint8Array([
+      (0b1000 << 4) | frame.opcode,
+      hasMask | 0b01111110,
+      payloadLength >>> 8,
+      payloadLength & 0x00ff
+    ]);
+  } else {
+    header = new Uint8Array([
+      (0b1000 << 4) | frame.opcode,
+      hasMask | 0b01111111,
+      ...sliceLongToBytes(payloadLength)
+    ]);
+  }
+  if (frame.mask) {
+    unmask(frame.payload, frame.mask);
+  }
+  const bytes = new Uint8Array(header.length + payloadLength);
+  bytes.set(header, 0);
+  bytes.set(frame.payload, header.length);
+  const w = new BufWriter(writer);
+  await w.write(bytes);
+  await w.flush();
+}
+
+export function unmask(payload: Uint8Array, mask: Uint8Array) {
+  if (mask) {
+    for (let i = 0; i < payload.length; i++) {
+      payload[i] ^= mask[i % 4];
+    }
+  }
+}
+
+export function acceptable(req: ServerRequest): boolean {
+  return (
+    req.headers.get("upgrade") === "websocket" &&
+    req.headers.has("sec-websocket-key")
+  );
+}
+
+export async function acceptWebSocket(req: ServerRequest): Promise<WebSocket> {
+  if (acceptable(req)) {
+    const sock = new WebSocketImpl(req.conn);
+    const secKey = req.headers.get("sec-websocket-key");
+    const secAccept = createSecAccept(secKey);
+    await req.respond({
+      status: 101,
+      headers: new Headers({
+        Upgrade: "websocket",
+        Connection: "Upgrade",
+        "Sec-WebSocket-Accept": secAccept
+      })
+    });
+    return sock;
+  }
+  throw new Error("request is not acceptable");
+}
+
+const kGUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+
+export function createSecAccept(nonce: string) {
+  const sha1 = new Sha1();
+  sha1.update(nonce + kGUID);
+  const bytes = sha1.digest();
+  const hash = bytes.reduce(
+    (data, byte) => data + String.fromCharCode(byte),
+    ""
+  );
+  return btoa(hash);
+}
+
+export async function readFrame(buf: BufReader): Promise<WebSocketFrame> {
+  let b = await buf.readByte();
+  let isLastFrame = false;
+  switch (b >>> 4) {
+    case 0b1000:
+      isLastFrame = true;
+      break;
+    case 0b0000:
+      isLastFrame = false;
+      break;
+    default:
+      throw new Error("invalid signature");
+  }
+  const opcode = b & 0x0f;
+  // has_mask & payload
+  b = await buf.readByte();
+  const hasMask = b >>> 7;
+  let payloadLength = b & 0b01111111;
+  if (payloadLength === 126) {
+    payloadLength = await readShort(buf);
+  } else if (payloadLength === 127) {
+    payloadLength = await readLong(buf);
+  }
+  // mask
+  let mask;
+  if (hasMask) {
+    mask = new Uint8Array(4);
+    await buf.readFull(mask);
+  }
+  // payload
+  const payload = new Uint8Array(payloadLength);
+  await buf.readFull(payload);
+  return {
+    isLastFrame,
+    opcode,
+    mask,
+    payload
+  };
+}
```

Comment:
I can't say I've gone through all of this in detail, but it looks good from a glance.